### PR TITLE
stdenv: handle $sourceRoot that starts with dash

### PR DESCRIPTION
**Pull request by trofi**
## Description of changes

Without the change the build fails on tarballs like
`diffoscope-269` that contain single top-level `-269` root as:

        diffoscope> unpacking source archive /nix/store/p620nidkm73vrp0z6kk5krmrm4vg7bxd-diffoscope-269.tar.bz2
        diffoscope> source root is -269
        diffoscope> setting SOURCE_DATE_EPOCH to timestamp 1717143039 of file ./-269/tests/utils/versions.py
        diffoscope> /nix/store/n7wnqwm52s5d85dwcn9dkp7fx7wvq543-stdenv-linux/setup: line 1589: cd: -2: invalid option
        diffoscope> cd: usage: cd [-L|[-P [-e]] [-@]] [dir]

Currently `diffoscope-269` has a `sourceRoot = "./-269";` workaround to
bypass the failure.

With the changes above the workaround is not needed anymore. Contains 4 changes:

 - ensureNewerSourcesHook: handle $sourceRoot that starts with dash
 - setup-hooks/set-source-date-epoch-to-latest.sh: handle $path that starts with dash
 - generic/setup.sh: handle $sourceRoot that starts with dash
 - generic/setup.sh: handle $sourceRoot that starts with dash

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by bjornfor** [comment]
Squash the two `generic/setup.sh: handle $sourceRoot that starts with dash` commits?

**Comment by SuperSandro2000** [comment]
Does find not support -- ?

**Comment by tie** [comment]
It doesn’t 😅

**Comment by trofi** [comment]
> Squash the two `generic/setup.sh: handle $sourceRoot that starts with dash` commits?

Squashed both.

File: pkgs/build-support/setup-hooks/set-source-date-epoch-to-latest.sh [hunk 1]
```diff
@@ -1,5 +1,9 @@
 updateSourceDateEpoch() {
     local path="$1"
+    # Avoid passing option-looking directory to find. The example
+    # is diffoscope-269:
```
Comment (philiptaron) [inline pkgs/build-support/setup-hooks/set-source-date-epoch-to-latest.sh:4]: Stupid line-wrapping suggestion:

```suggestion
    # Avoid passing option-looking directory to find. The example is diffoscope-269:
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -217,7 +217,12 @@ with pkgs;
   } (writeScript "ensure-newer-sources-hook.sh" ''
       postUnpackHooks+=(_ensureNewerSources)
       _ensureNewerSources() {
-        '${findutils}/bin/find' "$sourceRoot" \
+        local r=$sourceRoot
+        # Avoid passing option-looking directory to find. The example
+        # is diffoscope-269:
```
Comment (philiptaron) [inline pkgs/top-level/all-packages.nix:222]:
```suggestion
        # Avoid passing option-looking directory to find. The example is diffoscope-269:
```

**Review by philiptaron — APPROVED** [review]
Works for me. Couple of in-no-way-blocking suggestions for line wrapping in comments.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (trofi) [reply pkgs/top-level/all-packages.nix]: Merged manually.

**Review by philiptaron — APPROVED** [review]
Love it. Thanks.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -217,7 +217,11 @@ with pkgs;
   } (writeScript "ensure-newer-sources-hook.sh" ''
       postUnpackHooks+=(_ensureNewerSources)
       _ensureNewerSources() {
-        '${findutils}/bin/find' "$sourceRoot" \
+        local r=$sourceRoot
+        # Avoid passing option-looking directory to find. The example is diffoscope-269:
+        #   https://salsa.debian.org/reproducible-builds/diffoscope/-/issues/378
+        [[ $r == -* ]] && r="./$r"
+        '${findutils}/bin/find' "$r" \
```
Comment (tie) [inline pkgs/top-level/all-packages.nix:224]: Isn’t `find` part of the stdenv?

**Review by tie — APPROVED** [review]
LGTM (see also comment)

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (trofi) [reply pkgs/top-level/all-packages.nix:224]: I did not change the `find` source location in this patch. But looking at https://pubs.opengroup.org/onlinepubs/9699919799/utilities/find.html `POSIX` does not document `-newermt` option. Thus something needs to make sure it's `GNU findutils`.  The hook seems to apply to handle only `.zip` files, thus maybe it's not that important WRT closure size and extra dependencies.

Yes, today `findutils` is currently pulled into `stdenv` via `pkgs/stdenv/generic/common-path.nix`. I'm not sure if `stdenv` is guaranteed to have `GNU find` implementation (or any `find` implementation) in future.

**Review by rhendric — APPROVED** [review]

**Pull request merged**